File: exercises/challenges/easy_1/octal.rb
# Octal
#
# Implement octal to decimal conversion. Given an octal input string, your program should produce a decimal output.
#
# Note:
#
# Implement the conversion yourself. Do not use something else to perform the conversion for you. Treat invalid input as octal 0.
#
# About Octal (Base-8):
#
# Decimal is a base-10 system.
#
# A number 233 in base 10 notation can be understood as a linear combination of powers of 10:
#
#     The rightmost digit gets multiplied by 100 = 1
#     The next number gets multiplied by 101 = 10
#     ...
#     The n*th number gets multiplied by 10n-1*.
#     All these values are summed.
#
# So:
#
#   233 # decimal
# = 2*10^2 + 3*10^1 + 3*10^0
# = 2*100  + 3*10   + 3*1
#
# Octal is similar, but uses powers of 8 rather than powers of 10. So:
#
#   233 # octal
# = 2*8^2 + 3*8^1 + 3*8^0
# = 2*64  + 3*8   + 3*1
# = 128   + 24    + 3
# = 155

# Input: An octal number
# Output: an integer in decimal
#
# Return 0 if input is not an integer
# Or just convert any input to_i
#
# If we flip digits around, we can use index as exponent.

# Algorithm
# 0. Return 0 if number contains 8 or 9, since they dont exist in octal.
# 1. Flip digits in number
# 2. Convert digits to an array
# 3. Loop through array with index
# 4.   Multiply digit times 8^index
# 5. Add numbers in array together
# 6. Return sum

class Octal
  def initialize(value)
    @value = value.match?(/[89]/) ? 0 : value.to_i
  end

  def to_decimal
    @value.digits.map.with_index do |current_digit, index|
      current_digit * (8 ** index)
    end.sum
  end
end

File: lessons/215_lessons/lesson_1/octal.js
// Practice Problem: Octal
// Write a Function named octalToDecimal that performs octal to decimal conversion. When invoked on a String that contains the representation of an octal number, the Function returns a decimal version of that value as a Number. Implement the conversion yourself: do not use something else to perform the conversion for you.

// Input: A number as a string
// Output: A number of the Number type

// Rules:
// - implement the conversion yourself
// - input number is a string, string represents a number in octal
// - to convert a number: multiply digit by 8**place, from right to left, and exponent starts at 0
// - we can split number into an array

// Algorithm
// 1. Split string into an array
// 2. Reverse array
// 3. Map array
//  Turn element into a number
//  Return value for each element is element * 8**(current index)
// 4. Reduce array to one number

function octalToDecimal(numberString) {
  let number = numberString.split('').reverse();

  return number.map(function(digit, index) {
                  return Number(digit) * 8**index; })
               .reduce(function(sum, element) {
                  return sum + element;
               });
}

console.log(octalToDecimal('1'));           // 1
console.log(octalToDecimal('10'));          // 8
console.log(octalToDecimal('130'));         // 88
console.log(octalToDecimal('17'));          // 15
console.log(octalToDecimal('2047'));        // 1063
console.log(octalToDecimal('011'));         // 9
